diff --git a/cmd/mo-service/config.go b/cmd/mo-service/config.go
index 033d61b3f..56e63ddb9 100644
--- a/cmd/mo-service/config.go
+++ b/cmd/mo-service/config.go
@@ -462,6 +462,9 @@ func (c *Config) getCNServiceConfig() cnservice.Config {
 	cfg := c.CN
 	cfg.HAKeeper.ClientConfig = c.HAKeeperClient
 	cfg.Frontend.SetLogAndVersion(&c.Log, version.Version)
+	if cfg.Txn.Trace.Dir == "" {
+		cfg.Txn.Trace.Dir = "trace"
+	}
 	return cfg
 }
 
diff --git a/cmd/mo-service/dynamic.go b/cmd/mo-service/dynamic.go
index ccdf90f13..f4d97bece 100644
--- a/cmd/mo-service/dynamic.go
+++ b/cmd/mo-service/dynamic.go
@@ -124,6 +124,7 @@ func genDynamicCNConfigs(
 			uuid,
 			port,
 			i,
+			i,
 			frontendPort,
 			unixSocketPort)
 		f, err := os.CreateTemp(
diff --git a/cmd/mo-service/main.go b/cmd/mo-service/main.go
index 3752e5032..ea612e773 100644
--- a/cmd/mo-service/main.go
+++ b/cmd/mo-service/main.go
@@ -249,7 +249,7 @@ func startCNService(
 			cnservice.WithLogger(logutil.GetGlobalLogger().Named("cn-service").With(zap.String("uuid", cfg.CN.UUID))),
 			cnservice.WithMessageHandle(compile.CnServerMessageHandler),
 			cnservice.WithConfigData(commonConfigKVMap),
-			cnservice.WithTxnTraceData(filepath.Join(cfg.DataDir, "trace")),
+			cnservice.WithTxnTraceData(filepath.Join(cfg.DataDir, cfg.CN.Txn.Trace.Dir)),
 		)
 		if err != nil {
 			panic(err)
diff --git a/etc/launch-dynamic-cn/cn.toml.base b/etc/launch-dynamic-cn/cn.toml.base
index b1ed961dc..29debd7e9 100644
--- a/etc/launch-dynamic-cn/cn.toml.base
+++ b/etc/launch-dynamic-cn/cn.toml.base
@@ -38,6 +38,9 @@ uuid = "dd1dccb%d-4d3c-41f8-b482-5251dc7a41bf"
 port-base = %d
 service-host = "cn%d"
 
+[cn.txn.trace]
+dir = "trace%d"
+
 [cn.Engine]
 type = "distributed-tae"
 
diff --git a/pkg/cnservice/types.go b/pkg/cnservice/types.go
index f6d5f590e..bf86b8e91 100644
--- a/pkg/cnservice/types.go
+++ b/pkg/cnservice/types.go
@@ -240,6 +240,7 @@ type Config struct {
 			BufferSize    int           `toml:"buffer-size"`
 			FlushBytes    toml.ByteSize `toml:"flush-bytes"`
 			FlushDuration toml.Duration `toml:"force-flush-duration"`
+			Dir           string        `toml:"dir"`
 		} `toml:"trace"`
 	} `toml:"txn"`
 
